### PR TITLE
[rtl] Implement mvendorid/marchid/mimpid CSRs

### PR DESCRIPTION
**Pull request by GregAC**

### CODE REVIEW

File: doc/03_reference/cs_registers.rst
Comment (imphil) [inline doc/03_reference/cs_registers.rst:601]: Maybe add a note that these numbers are assigned to us, and people should change them if they make significant changes (can we be more specific?) to the core?

File: rtl/ibex_cs_registers.sv
Comment (imphil) [inline rtl/ibex_cs_registers.sv:129]: Ibex is commercial. I think we should just get a JEDEC ID, we'll need that one day for JTAG and similar purposes as well.

@asb

**Review by imphil — COMMENTED** [review]
Thanks Greg! 

I'd be interested in adding a bit more guidance to users of Ibex what they should and should not do with those IDs. E.g. when to change them (not fully detailed, but at least roughly). And if we want to encourage changing the IDs, maybe putting them in a place that requires less RTL modifications (such as a package) or even a top-level parameter?

**Comment by GregAC** [comment]
> I'd be interested in adding a bit more guidance to users of Ibex what they should and should not do with those IDs

Given what these are for is part of the RISC-V specification I was trying to avoid saying too much about them in the documentation as we don't just want to repeat the specification. I could add something saying look at the specific for further details.

> And if we want to encourage changing the IDs, maybe putting them in a place that requires less RTL modifications (such as a package) or even a top-level parameter?

I can move them to the top-level package. Trying to avoid drowning Ibex in too many top-level parameters and these aren't things you're likely to want to change on an instantiation by instantiation basis.

File: rtl/ibex_cs_registers.sv
Comment (GregAC) [reply rtl/ibex_cs_registers.sv:129]: We can certainly look into it, could switch this comment to say 'No JEDC ID has been allocated to lowRISC so value is 0'?

Comment (rswarbrick) [reply rtl/ibex_cs_registers.sv:129]: Sounds like a good formulation to me.

Comment (imphil) [reply rtl/ibex_cs_registers.sv]: Yeah, let's do that for the moment. And it's "JEDEC"

**Comment by imphil** [comment]
> Given what these are for is part of the RISC-V specification I was trying to avoid saying too much about them in the documentation as we don't just want to repeat the specification.

Yes, but nobody ever reads the spec :smile:  So if we can give people a hint in the "how to use ibex for your project" section of the docs (https://ibex-core.readthedocs.io/en/latest/02_user/getting_started.html) that they should maybe change mimpid if they make small changes to Ibex, and mvendorid/marchid when they make massive changes to it, but leave the identifiers as they are if they take ibex as-is?

**Review by imphil — APPROVED** [review]

File: rtl/ibex_pkg.sv
Comment (GregAC) [inline rtl/ibex_pkg.sv:579]: I am now wondering if we should use some of these bits to indicate Ibex configuration, e.g. if writeback stage is enabled. We could have one bit per config option but this could add up to a lot of bits. We could reserve some number of bits for a 'config ID' and we give configurations of interest a config ID (e.g. the configurations listed in the README would each have their own config ID).

We could also/in addition implement an Ibex specific CSR that gives detailed config information.

**Comment by GregAC** [comment]
@imphil I've added a paragraph to the docs in getting started

File: doc/02_user/getting_started.rst
Comment (imphil) [inline doc/02_user/getting_started.rst:18]:
```suggestion
The RISC-V Privileged Architecture specifies several read-only CSRs that identify the vendor and micro-architecture of a CPU.
```

micro-architecture for consistency with the usage below.

File: doc/02_user/getting_started.rst
Comment (imphil) [inline doc/02_user/getting_started.rst:22]:
```suggestion
Ibex, as an open source implementation, has an assigned architecture ID (``marchid``) of 22. 
(Allocations are specified in `marchid.md of the riscv-isa-manual repository <https://github.com/riscv/riscv-isa-manual/blob/master/marchid.md>`_.)
```

File: doc/03_reference/cs_registers.rst
Comment (imphil) [inline doc/03_reference/cs_registers.rst:593]:
```suggestion
Details of what the ID represents can be found in the RISC-V Privileged Specification.
```

(We specify the version of the RISC-V spec that we follow in one place to avoid having to update them everywhere if we change it.)

Also below.

**Review by imphil — APPROVED** [review]

**Comment by vogelpi** [comment]
Hey @GregAC ,
it has been a while since the last activity in this PR. But it's already approved and if I understand correctly, the changes are absolutely uncontroversial. We could just rebase and merge this PR. Or is there is there still something missing?

**Comment by GregAC** [comment]
Ah well spotted, think I simply forgot to hit the merge button! There's a potential add-on in using mimpid to differentiate different Ibex configs but can deal with that in a follow up if desired.

**Comment by vogelpi** [comment]
Ah, great. Thanks for the quick feedback Greg.

**Pull request merged**
